go/service/saltpack.go

```diff
@@ -717,3 +868,110 @@ func (sf *sourceFile) Close() error {
 func (sf *sourceFile) reporter(bytesComplete, bytesTotal int64) {
 	sf.G().NotifyRouter.HandleSaltpackOperationProgress(context.Background(), sf.op, sf.filename, bytesComplete, bytesTotal)
 }
+
+func isDir(filename string) (bool, error) {
+	s, err := os.Stat(filename)
+	if err != nil {
+		return false, err
+	}
+	return s.IsDir(), nil
+}
+
+func dirTotalSize(filename string) (int64, error) {
+	var total int64
+	err := filepath.Walk(filename, func(path string, info os.FileInfo, inErr error) error {
+		if inErr != nil {
+			return inErr
+		}
+		if info.IsDir() {
+			return nil
+		}
+		total += info.Size()
+		return nil
+	})
+	return total, err
+}
+
+// zipDir will create a zip archive of everything in directory suitablie for streaming.
+// It will send something on the chan error when it is done.
+// Read the data of the zip archive from the returned io.ReadCloser.
+func zipDir(directory string, prog *progress.ProgressWriter) (chan error, io.ReadCloser) {
+	// make a pipe so we can give saltpack engines a reader for the zip source
+	pipeRead, pipeWrite := io.Pipe()
+
+	// connect a zip archiver to the writer side of the pipe
+	w := zip.NewWriter(pipeWrite)
+
+	// given a directory like "a/b/c/d" or "/opt/blah/d", we want the files in the
+	// zip file to recreate just the final directory.  In other words, d/, d/1/, d/1/000.log
+	// and *not* a/b/c/d/, a/b/c/d/1/, a/b/c/d/1/000.log.
+	parent := filepath.Dir(directory)
+	stripParent := func(s string) string {
+		return strings.TrimPrefix(s, parent+string(filepath.Separator))
+	}
+
+	zipch := make(chan error)
+	go func() {
+		err := filepath.Walk(directory, func(path string, info os.FileInfo, inErr error) error {
+			if inErr != nil {
+				return inErr
+			}
+
+			stripped := stripParent(path)
+
+			header, err := zip.FileInfoHeader(info)
+			if err != nil {
+				return err
+			}
+			header.Method = zip.Deflate
+
+			if info.IsDir() {
+				// make a directory by calling Create with a filename ending in a separator.
+				header.Name = stripped + string(filepath.Separator)
+				_, err := w.CreateHeader(header)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+
+			// make a regular file and copy all the data into it.
+			header.Name = stripped
+			w, err := w.CreateHeader(header)
+			if err != nil {
+				return err
+			}
+			f, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			tr := io.TeeReader(bufio.NewReader(f), prog)
+			_, err = io.Copy(w, tr)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+
+		// close everything
+		w.Close()
+		pipeWrite.Close()
```

Comment:
Check error values from Close. I don't think pipes can error, just from the code, though didn't see docs. Zip writer can error on Close.